resource/aws_fsx_windows_file_system: Unexpected errors and panics with capacity updates

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Terraform CLI and Terraform AWS Provider Version

```
Terraform CLI v0.12.29

Terraform AWS Provider main branch
```

### Affected Resource(s)

* aws_fsx_windows_file_system

### Terraform Configuration Files

```hcl
variable "storage_capacity" {
  type = number
}

resource "aws_fsx_windows_file_system" "test" {
  active_directory_id = aws_directory_service_directory.test.id
  skip_final_backup   = true
  storage_capacity    = var.storage_capacity
  subnet_ids          = [aws_subnet.test1.id]
  throughput_capacity = 16
}
```

### Debug Output

Unexpected error:

```
2020/11/24 17:55:04 [DEBUG] [aws-sdk-go] DEBUG: Response FSx/DescribeFileSystems Details:
---[ RESPONSE ]--------------------------------------
HTTP/1.1 200 OK
Connection: close
Content-Length: 1250
Content-Type: application/x-amz-json-1.1
Date: Tue, 24 Nov 2020 17:55:04 GMT
X-Amzn-Requestid: e15748d3-3074-48a0-ad74-9aae0e00d4a1


-----------------------------------------------------
2020/11/24 17:55:04 [DEBUG] [aws-sdk-go] {"FileSystems":[{"AdministrativeActions":[{"AdministrativeActionType":"FILE_SYSTEM_UPDATE","RequestTime":1.606240473954E9,"Status":"PENDING","TargetFileSystemValues":{"StorageCapacity":36}},{"AdministrativeActionType":"STORAGE_OPTIMIZATION","RequestTime":1.606240473954E9,"Status":"PENDING"}],"CreationTime":1.606240455065E9,"DNSName":"fs-05f1b32969f95e5dc.corp.notexample.com","FileSystemId":"fs-05f1b32969f95e5dc","FileSystemType":"WINDOWS","KmsKeyId":"arn:aws:kms:us-west-2:--OMITTED--:key/21d2faae-2124-4cb5-9636-0159633a9f08","Lifecycle":"AVAILABLE","NetworkInterfaceIds":["eni-076385c73f83789e5"],"OwnerId":"187416307283","ResourceARN":"arn:aws:fsx:us-west-2:--OMITTED--:file-system/fs-05f1b32969f95e5dc","StorageCapacity":32,"StorageType":"SSD","SubnetIds":["subnet-09422bb0cd6ba51a7"],"Tags":[],"VpcId":"vpc-0644e8d2d1f523693","WindowsConfiguration":{"ActiveDirectoryId":"d-9267082361","AutomaticBackupRetentionDays":7,"CopyTagsToBackups":false,"DailyAutomaticBackupStartTime":"13:30","DeploymentType":"SINGLE_AZ_1","PreferredFileServerIp":"10.0.1.254","PreferredSubnetId":"subnet-09422bb0cd6ba51a7","RemoteAdministrationEndpoint":"fs-05f1b32969f95e5dc.corp.notexample.com","ThroughputCapacity":16,"WeeklyMaintenanceStartTime":"6:07:00"}}]}
2020/11/24 17:55:04 [WARN] Got error running Terraform: 2020/11/24 17:54:32 [DEBUG] Using modified User-Agent: Terraform/0.12.29 HashiCorp-terraform-exec/0.10.0

Error: Error waiting for filesystem (fs-05f1b32969f95e5dc) to become available: unexpected state 'PENDING', wanted target 'COMPLETED, UPDATED_OPTIMIZING'. last error: %!s(<nil>)
```

Panic:

```
2020/11/24 17:50:12 [DEBUG] [aws-sdk-go] {"FileSystems":[{"CreationTime":1.606240165654E9,"DNSName":"amznfsxnll6tba2.corp.notexample.com","FileSystemId":"fs-0e21d63b1fa7770d9","FileSystemType":"WINDOWS","KmsKeyId":"arn:aws:kms:us-west-2:--OMITTED--:key/21d2faae-2124-4cb5-9636-0159633a9f08","Lifecycle":"UPDATING","NetworkInterfaceIds":["eni-0082d3d290934910d"],"OwnerId":"187416307283","ResourceARN":"arn:aws:fsx:us-west-2:--OMITTED--:file-system/fs-0e21d63b1fa7770d9","StorageCapacity":32,"StorageType":"SSD","SubnetIds":["subnet-08771b7bc84d10514"],"Tags":[],"VpcId":"vpc-03d8a668635aebc2a","WindowsConfiguration":{"AutomaticBackupRetentionDays":7,"CopyTagsToBackups":false,"DailyAutomaticBackupStartTime":"11:30","DeploymentType":"SINGLE_AZ_1","PreferredFileServerIp":"10.0.1.237","PreferredSubnetId":"subnet-08771b7bc84d10514","RemoteAdministrationEndpoint":"amznfsxnll6tba2.corp.notexample.com","SelfManagedActiveDirectoryConfiguration":{"DnsIps":["10.0.1.105","10.0.2.43"],"DomainName":"corp.notexample.com","FileSystemAdministratorsGroup":"Domain Admins","UserName":"Administrator"},"ThroughputCapacity":8,"WeeklyMaintenanceStartTime":"7:11:30"}}]}
```

### Panic Output

```
=== CONT  TestAccAWSFsxWindowsFileSystem_SelfManagedActiveDirectory_Username
------- Stderr: -------
panic: runtime error: index out of range [0] with length 0
goroutine 5098 [running]:
github.com/terraform-providers/terraform-provider-aws/aws.refreshFsxFileSystemLifecycleOptimizing.func1(0xc00011def0, 0xc00011de48, 0x2, 0x1, 0x1, 0xc000512580)
  /opt/teamcity-agent/work/2e10e023da0c7520/src/github.com/terraform-providers/terraform-provider-aws/aws/fsx.go:67 +0x176
github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource.(*StateChangeConf).WaitForStateContext.func1(0xc000b48180, 0xc0014b0230, 0xc0012eca20, 0xc0021a1840, 0xc0010cf150, 0xc0010cf148)
  /opt/teamcity-agent/work/2e10e023da0c7520/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.3.0/helper/resource/state.go:110 +0x329
created by github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource.(*StateChangeConf).WaitForStateContext
  /opt/teamcity-agent/work/2e10e023da0c7520/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.3.0/helper/resource/state.go:83 +0x1c5
```

Other tests similarly panicked:

```
=== CONT  TestAccAWSFsxWindowsFileSystem_WeeklyMaintenanceStartTime
------- Stderr: -------
panic: runtime error: index out of range [0] with length 0

=== CONT  TestAccAWSFsxWindowsFileSystem_DailyAutomaticBackupStartTime
------- Stderr: -------
panic: runtime error: index out of range [0] with length 0

=== CONT  TestAccAWSFsxWindowsFileSystem_AutomaticBackupRetentionDays
------- Stderr: -------
panic: runtime error: index out of range [0] with length 0
```

### Expected Behavior

`PENDING` status should be handled in waiter. No panics. 😄 

### Actual Behavior

`PENDING` status errors unexpectedly. Missing `AdministrativeActions` array panics.

### Steps to Reproduce

1. `TF_ACC=1 go test -parallel=20 -run='TestAccAWSFsxWindowsFileSystem_' -timeout=4h -v`

### References

* Code recently introduced: https://github.com/hashicorp/terraform-provider-aws/pull/15582

--- a/aws/fsx.go
+++ b/aws/fsx.go
@@ -48,7 +48,7 @@ func refreshFsxFileSystemLifecycle(conn *fsx.FSx, id string) resource.StateRefre
 	}
 }
 
-func refreshFsxFileSystemLifecycleOptimizing(conn *fsx.FSx, id string) resource.StateRefreshFunc {
+func refreshFsxFileSystemAdministrativeActionsStatusFileSystemUpdate(conn *fsx.FSx, id string) resource.StateRefreshFunc {
 	return func() (interface{}, string, error) {
 		filesystem, err := describeFsxFileSystem(conn, id)
 
@@ -64,7 +64,17 @@ func refreshFsxFileSystemLifecycleOptimizing(conn *fsx.FSx, id string) resource.
 			return nil, "", nil
 		}
 
-		return filesystem, aws.StringValue(filesystem.AdministrativeActions[0].Status), nil
+		for _, administrativeAction := range filesystem.AdministrativeActions {
+			if administrativeAction == nil {
+				continue
+			}
+
+			if aws.StringValue(administrativeAction.AdministrativeActionType) == fsx.AdministrativeActionTypeFileSystemUpdate {
+				return filesystem, aws.StringValue(administrativeAction.Status), nil
+			}
+		}
+
+		return filesystem, fsx.StatusCompleted, nil
 	}
 }
 
@@ -110,14 +120,17 @@ func waitForFsxFileSystemUpdate(conn *fsx.FSx, id string, timeout time.Duration)
 	return err
 }
 
-func waitForFsxFileSystemUpdateOptimizing(conn *fsx.FSx, id string, timeout time.Duration) error {
+func waitForFsxFileSystemUpdateAdministrativeActionsStatusFileSystemUpdate(conn *fsx.FSx, id string, timeout time.Duration) error {
 	stateConf := &resource.StateChangeConf{
-		Pending: []string{fsx.StatusInProgress},
+		Pending: []string{
+			fsx.StatusInProgress,
+			fsx.StatusPending,
+		},
 		Target: []string{
 			fsx.StatusCompleted,
 			fsx.StatusUpdatedOptimizing,
 		},
-		Refresh: refreshFsxFileSystemLifecycleOptimizing(conn, id),
+		Refresh: refreshFsxFileSystemAdministrativeActionsStatusFileSystemUpdate(conn, id),
 		Timeout: timeout,
 		Delay:   30 * time.Second,
 	}

--- a/aws/resource_aws_fsx_windows_file_system.go
+++ b/aws/resource_aws_fsx_windows_file_system.go
@@ -31,6 +31,7 @@ func resourceAwsFsxWindowsFileSystem() *schema.Resource {
 		Timeouts: &schema.ResourceTimeout{
 			Create: schema.DefaultTimeout(45 * time.Minute),
 			Delete: schema.DefaultTimeout(30 * time.Minute),
+			Update: schema.DefaultTimeout(45 * time.Minute),
 		},
 
 		Schema: map[string]*schema.Schema{
@@ -334,12 +335,13 @@ func resourceAwsFsxWindowsFileSystemUpdate(d *schema.ResourceData, meta interfac
 
 	if requestUpdate {
 		_, err := conn.UpdateFileSystem(input)
+
 		if err != nil {
-			return fmt.Errorf("error updating FSX File System (%s): %s", d.Id(), err)
+			return fmt.Errorf("error updating FSx Windows File System (%s): %w", d.Id(), err)
 		}
 
-		if err := waitForFsxFileSystemUpdateOptimizing(conn, d.Id(), d.Timeout(schema.TimeoutCreate)); err != nil {
-			return fmt.Errorf("Error waiting for filesystem (%s) to become available: %w", d.Id(), err)
+		if err := waitForFsxFileSystemUpdateAdministrativeActionsStatusFileSystemUpdate(conn, d.Id(), d.Timeout(schema.TimeoutUpdate)); err != nil {
+			return fmt.Errorf("error waiting for FSx Windows File System (%s) update: %w", d.Id(), err)
 		}
 	}
 

--- a/website/docs/r/fsx_windows_file_system.html.markdown
+++ b/website/docs/r/fsx_windows_file_system.html.markdown
@@ -100,6 +100,7 @@ configuration options:
 
 * `create` - (Default `45m`) How long to wait for the file system to be created.
 * `delete` - (Default `30m`) How long to wait for the file system to be deleted.
+* `update` - (Default `45m`) How long to wait for the file system to be updated.
 
 ## Import
 

